### ISSUE TITLE
No way to check if modal has been shown

### ISSUE THREAD

**Issue by webframing** [ng-bootstrap/ng-bootstrap#3039]
Though we have the options to do something when the modal has been closed by the help of the Promise returned by the `NgbModal.open(template).result`, but I searched [here](https://ng-bootstrap.github.io/#/components/modal/examples) in the documentation for a way to check whether the modal has been opened, I didn't find.

Actually, the core library of the bootstrap provides it by triggering an event:-

```javascript
$('#myModal').on('shown.bs.modal', function (e) {
  // do something...
})
```
I need the above to create a scrollable modal that allows scrolling the modal body by adding `.modal-dialog-scrollable` to `.modal-dialog`. (That is also not supported by the ng-bootstrap yet).
You can see [here](https://stackblitz.com/edit/ng-modal-demo) on the stackblitz or in the code below. I have to wait for one second so that we can have the assurance that the modal has been opened.

```javascript
// open modal
  openModal() {
    this._NgbModal.open(NgModalComponent, {
      windowClass: 'modal-job-scrollable'
    });

    // upwrap the "app-ng-modal" data to enable the "modal-dialog-scrollable"
    // and make the modal scrollable
    (() => {
      const node: HTMLElement | null = document.querySelector('app-ng-modal');
      if (node) {
        while (node.firstChild) {
          (node.parentNode as HTMLElement).insertBefore(node.firstChild, node);
        }
      }
      // make the modal scrollable by adding the class .modal-dialog-scrollable
      // here wait for one second so that we can find the .modal-dialog
      setTimeout(() => {
        const modalDialog = document.querySelector('.modal-job-scrollable .modal-dialog');
        if (modalDialog) {
          modalDialog.classList.add('modal-dialog-scrollable');
        }
      }, 1000)
    })();
  }
```
Here I am using [Components as content](https://ng-bootstrap.github.io/#/components/modal/examples#component) appraoch of the ng-bootstrap.

**Comment by maxokorokov** [comment]
Please note, that scrollable option will be released in #2909 

As it was mentioned in https://github.com/ng-bootstrap/ng-bootstrap/issues/3005#issuecomment-461348714, the modal `.open()` is synchronous. Let me provide **more** details, so here is the code. Going to leave it as a reference:

```ts
open(content) {
    this.modalService.open(...);

    // after .open()

    this._zone.onStable.pipe(take(1)).subscribe(() => {
      // zone on stable
    });

    requestAnimationFrame(() => {
      // RAF
    });

    setTimeout(() => {
      // timeout
    });
  }
```

Here what it would look like in the timeline with four marks: `after .open()`, on `zone stable`, on `RAF (request animation frame` and on `setTimeout()`:

![screen shot 2019-02-25 at 14 20 01](https://user-images.githubusercontent.com/8074436/53341005-ac9df500-390a-11e9-91a0-77abcb58a043.png)

So, (1) when we **synchronously** create the modal, we create just some DOM nodes, so `after .open()` you'll have something like this:

![screen shot 2019-02-25 at 14 09 09](https://user-images.githubusercontent.com/8074436/53340578-a9563980-3909-11e9-941f-d543919b9f73.png)

Note that Angular change detection is executed a bit later when leaving the zone. RIght after change detection you'll have something like this:

![screen shot 2019-02-25 at 14 09 58](https://user-images.githubusercontent.com/8074436/53340665-dd315f00-3909-11e9-93c9-2e83d0359814.png)

The earliest you can be this is in (2) `zone.onStable` or in the `requestAnimationFrame` depending on your needs.

If you need the callback after layout and paint happened → use (3) `setTimeout()`

--- 

So:
- you can be notified in any of these points in time
- What does it mean to be opened? We have NO idea what is in the content of the modal. It might render synchronously and be available on zone stable, it might not. So it's not very clear WHEN the `opened` event should be raised.

Concerning your code:
- you probably shouldn't wait for 1 second, but use one of the callbacks.
- here is a better stackblitz: https://stackblitz.com/edit/ng-modal-demo-qcaqaa?file=src/app/app.component.ts

---

Will mark it as a feature request and leave open.

**Comment by webframing** [comment]
@maxokorokov Thanks...! Understood your nice explanation. Yes, probably I shouldn't wait for 1 second. For now using the method provided in the better stackblitz, and waiting for the next release.

By the way, most of the modal customizations are achieved via applying classes on `.modal-dialog`, then why ng-bootstrap does not provide one more option to apply the classes on the `.modal-dialog`. That can cover actually three options from the configuration:-

1. centered (using `.modal-dialog-centered`)
2. size (using `.modal-sm`, `.modal-lg` and `.modal-xl`)
3. scrollable (using `.modal-dialog-scrollable`)

**Comment by maxokorokov** [comment]
> then why ng-bootstrap does not provide one more option to apply the classes on the `.modal-dialog`

I think the issue where it was discussed/addressed is #818, but implementations were made only for the `backdropClass` and `windowClass`. 

Looks like a valid feature request to me (that should be handled separately form this one), but maybe I'm missing something.

**Comment by maxokorokov** [comment]
- classes can be applied now to `.modal-dialog`
- there are `shown` and `hidden` options now with animations since `v8.0.0` → https://ng-bootstrap.github.io/#/components/modal/api#NgbModalRef

Closing this
